README.md
# Decimal JS SDK

For detailed explanation on how things work, checkout the:

- [Decimal SDK docs](https://help.decimalchain.com/api-sdk/).
- [Decimal Console site](https://console.decimalchain.com/).

# Install

```bash
$ npm install decimal-js-sdk
```

# Usage

## Connecting

You can provide your own Decimal node address to send transactions to the network, or you can use our gateways to do this and much more. Gateway URLs are provided below.

MAINNET -
https://mainnet-gate.decimalchain.com/api/

TESTNET -
https://testnet-gate.decimalchain.com/api/

Depending on what you wish to use provide either gateUrl or restURL (your node rest service url) in config object passed to the Decimal instance

```js

const decimalOptions = {
  gateUrl: 'https://testnet-gate.decimalchain.com/api/',
  // restURL: 'https://your-node.example.com/rest' if you want to use your own Decimal node instance
}
// baseURL option used in old versions is now deprecated
const decimal = new Decimal(decimalOptions);
```

The following code can be used to generate **mnemonic** or use your mnemonic

```js
const bip39 = require("bip39");
const mnemonic = bip39.generateMnemonic();
```

### Get generated user  wallets from DB

```js
// add oprions to wallet object and call method synchronize
const options = {
  gateUrl: 'https://testnet-gate.decimalchain.com/api/',
}

const wallet = new Wallet(/*your mnemonic*/, options)

// get saved wallet ids from DB

wallet.getAndUseGeneratedWallets()

// save new generated wallet in DB
// before save it, use method wallet.getAndUseGeneratedWallets()

wallet.sendAndSaveGeneratedWallets()
```


## Send coins

### CommonJS

```js
const { Wallet, Decimal } = require('decimal-js-sdk'); // For server use 'decimal-js-sdk/dist/decimal-sdk-node'

const wallet = new Wallet( /*your mnemonic*/);


const decimalOptions = {
  restURL: 'https://your-node.example.com/rest' if you want to use your own Decimal node instance
}
// baseURL option used in old versions is now deprecated
const decimal = new Decimal(decimalOptions);

const data = {
  to: 'dx13ykakvugqwzqqmqdj2j2hgqauxmftdn3kqy69g',
  coin: 'tdel',
  amount: '1',
};

const txResult = await decimal.sendCoins(data, [options]); // see options for details
// => {hash: '4C0A408B6EBC33AD...', success: true, error: null}
```

### ES6

```js

import { Wallet, Decimal } from 'decimal-js-sdk';

const wallet = new Wallet(/*your mnemonic*/);
const decimal = new Decimal({gateUrl: 'https://testnet-gate.decimalchain.com/api/', wallet})

const data = {
  to: 'dx13ykakvugqwzqqmqdj2j2hgqauxmftdn3kqy69g',
  coin: 'tdel',
  amount: '1',
};

const txResult = await decimal.sendCoins(data); // see sendCoins for details
// => {hash: '4C0A408B6EBC33AD...', success: true, error: null}
```

## Wallet API

### Generate wallet

```js
import { Wallet } from 'decimal-js-sdk';

const wallet = new Wallet();
```

### Generate wallet from mnemonic

```js
import { Wallet } from 'decimal-js-sdk';

const wallet = new Wallet('erase august mask elevator sand picture north there apple equal anchor target');
```

### Wallet methods

`getPrivateKeyString`
Return the private key string of 64 hex characters

```js
wallet.getPrivateKeyString();
// 8f7370936a728ff3d2306cbf2422382a4297f9059aecb5703fe83eef49d3e828
```

`getPublicKeyString`
Return the public key string of 64 hex characters

```js
wallet.getPublicKeyString();
// 02f85bb9b5d87cda4197b09d0cd0ef5a1fb97f74f261f508f48a26d7dba2d0ed12
```

`switchAccount(id)`
Switches to wallet from list of wallets, generated with Master key, updates wallet to picked, wallet id picked, private key, public key, address. Only to generated wallets.

```js
wallet.switchAccount(id);
// switched to another early generated wallet.privateKey, wallet.publicKey, wallet.address
```

`generateAccount`
Generates new wallet with Master key, updates wallet to new generated, account's depth, wallet id to new generated, list of wallets, private key, public key, address. Maximum to 20 wallets with current Master key.

```js
wallet.generateAccount();
// new wallet.privateKey, wallet.publicKey, wallet.address
```

`generateAndSwitchAccount`
Generates new wallets with Master key and switches to wallet with provided id or the last wallet. Maximum to 20 wallets with current Master key.

```js
wallet.generateAndSwitchAccount(depth, id);
// new wallets to depth, switched to wallet with provided id
```

# SDK instance

- [Options](#options)
- [Types](#types)
- [Utilities](#Utilities)
  - [verifyAddress](#verifyAddress)
- [API](#api)
  - [getCoinsList](#getcoinslist)
  - [getCoin](#getcoin)
  - [getAddress](#getaddress)
  - [getMeta](#getmeta)
  - [getMultisigsByAddress](#getmultisigsbyaddress)
  - [getMultisig](#getmultisig)
  - [getMultisigTxs](#getmultisigtxs)
  - [getNft](#getNft)
  - [getNfts](#getNfts)
  - [getNftTxes](#getNftTxes)
  - [getNftsTxes](#getNftsTxes)
  - [getStakesByAddress](#getstakesbyaddress)
  - [getValidator](#getvalidator)
- [Tx utils](#txutils)
  - [getTransaction](#gettransaction)
  - [postTx](#posttx)
  - [estimateTxFee](#estimatetxfee)
  - [getTransactionByHash](#getTransactionByHash)
- [Coins](#coins)
  - [sendCoins](#sendCoins)
  - [burnCoins](#burnCoins)
  - [multisendCoins](#multisendcoins)
  - [buyCoins](#buycoins)
  - [sellCoins](#sellcoins)
  - [sellAllCoins](#sellallcoins)
  - [createCoin](#createcoin)
- [Validators](#validators)
  - [validatorDelegate](#validatordelegate)
  - [validatorUnbond](#validatorunbond)
  - [validatorDeclare](#validatordeclare)
  - [validatorEdit](#validatoredit)
  - [validatorOn](#validatoron)
  - [validatorOff](#validatoroff)
- [Checks](#nfts)
  - [nftMint](#nftmint)
  - [nftBurn](#nftburn)
  - [nftTransfer](#nfttransfer)
  - [nftDelegate](#nftdelegate)
  - [nftUnbond](#nftunbond)
  - [nftUpdateReserve](#nftupdatereserve)
- [Swap](#swap)
  - [swapInit](#swap-init)
  - [swapRedeem](#swap-redeem)
- [NFTs](#checks)
  - [issueCheck](#issuechek)
  - [redeemCheck](#redeemcheck)
- [Multisig](#multisig)
  - [multisigCreateWallet](multisigcreatewallet)
  - [multisigCreateTx](multisigcreatetx)
  - [multisigSignTx](multisigsigntx)

## Network

Additionally, you can pass network parameter to enable network specific features:

```js
const decimal = new Decimal({
  gateUrl: 'https://testnet-gate.decimalchain.com/api/',
  network: 'testnet', // Enable testnet only txs types
  wallet,
  meta
})
```

Available networks: ['mainnet', 'testnet', 'devnet']
If omitted defaults to 'mainnet', which is fine in most cases.

## Utilities

### verifyAddress

```js
decimal.verifyAddress(address, prefix = 'dx')
```
- examples:

```js
const decimal = new Decimal(...)

const isValid = await decimal.verifyAddress('dx13wv0w5w6uzkqceevtcefgc3sk0gg33vjuew2wy') // true

const isValidatorAddress = await decimal.verifyAddress('dxvaloper13wv0w5w6uzkqceevtcefgc3sk0gg33vjuew2wy', 'dxvaloper') // true

```

## Types

``` js
  import {TX_TYPE} from 'decimal-js-sdk'
```

* TX_TYPE.COIN_BUY
* TX_TYPE.COIN_CREATE
* TX_TYPE.COIN_SELL
* TX_TYPE.COIN_SEND
* TX_TYPE.COIN_MULTISEND
* TX_TYPE.COIN_SELL_ALL
* TX_TYPE.COIN_REDEEM_CHECK
* TX_TYPE.COIN_ISSUE_CHECK
* TX_TYPE.VALIDATOR_CANDIDATE
* TX_TYPE.VALIDATOR_DELEGATE
* TX_TYPE.VALIDATOR_SET_ONLINE
* TX_TYPE.VALIDATOR_SET_OFFLINE
* TX_TYPE.VALIDATOR_UNBOND
* TX_TYPE.VALIDATOR_CANDIDATE_EDIT
* TX_TYPE.MULTISIG_CREATE_WALLET
* TX_TYPE.MULTISIG_CREATE_TX
* TX_TYPE.MULTISIG_SIGN_TX

## Meta

Used for generating offline transactions.
**This is an optional parameter!**

```js
const meta = {
  account_number: "10",
  chain_id: "decimal-testnet-06-30-15-30",
  sequence: "198"
}
```

## Options

Each transaction method consists of input data (data) and options.
**This is an optional parameter!**

```js
const options = {
  gasLimit: '9000000000000000000', // The maximum amount of gas you are willing to pay for a transaction
  feeCoin: 'BTC', // The coin that pays commission
  feeAmount: '4000', // Fee pay amount
  message: 'my message', // Any additional information about the transaction
  txBroadcastMode: 'sync', // Broadcast tx mode {sync | async | block}
  accountInfoMode: 'blockchain-with-mempool', // Method of obtaining nonce, receive nonce with unconfirmed transactions in a mempool or not {blockchain | blockchain-with-mempool}
  setNonceAutomatically: true, // Automatically calculate nonce within a small transaction pool, after getting the first one from blockchain {true | false}
  nonce: '100', // Custom nonce for the transaction {valid number string}
};
```

## Tx utils

### getTransaction()

```js
import {TX_TYPE} from 'decimal-js-sdk';

const broadcastTx = await decimal.getTransaction(TX_TYPE.COIN_SEND, data, options);
/*
{
  txBroadcastMode: "sync"
  tx: {
    fee: {amount: Array(0), gas: "9000000000000000000"},
    memo: "sdk test",
    msg: [{…}],
    signatures: [{…}]
  }
}
*/

```

### postTx()

```js
import {TX_TYPE} from 'decimal-js-sdk';

const broadcastTx = await decimal.getTransaction(TX_TYPE.COIN_SEND, data, options);
const result = await decimal.postTx(broadcastTx);

/*
{
  error: null
  hash: "244AA816B881BDA43704867574114D905631EFA3B6AFD0E401C7B603D8264DF8"
  success: true
}
*/
```

### estimateTxFee()

```js
const fee = await decimal.estimateTxFee(TX_TYPE.COIN_SEND, data, options);
console.log(fee); // 0.44

```

### getTransactionByHash()

```js
await decimal.getTransactionByHash(hash);
```


## API

Please note, that these methods require gateUrl to be provided

### getCoinsList()

```js
const limit = 10;
const offset = 0;
const query = ''; // search
await decimal.getCoinsList(limit, offset, query);

/*
{
  "count": 5,
  "coins": [
    {
      "symbol": "mycoin",
      "title": "MYCOIN",
      "volume": "1023454326378015905553",
      "reserve": "12609116636032889229477",
      "crr": 10,
      "limitVolume": "2000000000000000000000",
      "creator": "dx1vkn4lje42gjmyghc3vneg0yqa39wfgqvh4f8zg",
      "txHash": "93AC6D86FF12BD9594889FEF4092542CBA6FEFA09F5E4A710D9E6CF12C1006A7",
      "blockId": 1529,
      "avatar": "data:image/png;base64...",
      "createdAt": "2020-07-28T14:09:38.852Z",
      "updatedAt": "2020-07-29T12:25:56.512Z"
    },
    // ...
  ]
}
*/
```

### getCoin()

```js
const coinTicker = 'MYCOIN'
await decimal.getCoin(coinTicker);

/*
{
  "symbol": "mycoin",
  "title": "MYCOIN",
  "volume": "1022451550064246238748",
  "reserve": "12486116326032889229477",
  "crr": 10,
  "limitVolume": "2000000000000000000000",
  "creator": "dx1vkn4lje42gjmyghc3vneg0yqa39wfgqvh4f8zg",
  "txHash": "93AC6D86FF12BD9594889FEF4092542CBA6FEFA09F5E4A710D9E6CF12C1006A7",
  "blockId": 1529,
  "avatar": "data:image/png;base64..."
  "createdAt": "2020-07-28T14:09:38.852Z",
  "updatedAt": "2020-07-29T08:09:44.203Z"
}
*/
```

### getAddress()

```js
const address = 'dx13ykakvugqwzqqmqdj2j2hgqauxmftdn3kqy69g' // address of requested user, required property
const txLimit = 10; // optional property

await decimal.getAddress(address, txLimit);

/*
{
  "address": {
    "id": 31,
    "address": "dx13ykakvugqwzqqmqdj2j2hgqauxmftdn3kqy69g",
    "type": "single",
    "balance": {
      "tdel": "14999120000000000000000"
    },
    "balanceNft": [
      {
        amount: "1"
        collection: "testovaya"
        // if user's address then original cover of nft
        // else a placeholder image used for common display
        cover: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAYI
        nftId: "4a482a53b16345bb625a214b2f9c3f2968461cd0"
        nftReserve: [{...}]
        nftStake: [{...}]
      }
    ],
    "nonce": 2,
    "txes": 5,
    "createdAt": "2020-07-29T07:48:15.089Z",
    "updatedAt": "2020-07-29T08:07:36.086Z",
    "totalStake": null,
    "txs": [
      // ...
    ]
  },
  "coins": [
    // ...
  ]
}
*/
```

### getMeta()

```js

CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

### v0.3.2 -  2022-08-03
#### Fixes
* Bugs with implementation of new unstable interface [`PR`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/57)

### v0.2.8 -  2022-07-11
#### Fixes
* Description of changelog change [`#33386`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/54)

### v0.2.7 -  2022-07-08
#### Features
* Update changelog [`#33386`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/51)

### v0.2.4 -  2022-06-30
#### Features
* Added changelog to the project [`#32815`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/51)

### v0.2.1 -  2022-06-27
#### Fixes
* Changed fee for burn coin [`#32299`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/50)

### v0.2.0 -  2022-06-27
#### Features
* New transaction type "burn coin" [`#32299`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/49)

### v0.1.133 -  2022-05-24
#### Features
* New way to automatically set nonce within a small transaction pool [`#29351`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/48)

### v0.1.132 -  2022-05-06
#### Features
* Pattern to check in redeem [`#26638`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/47)

### v0.1.131 -  2022-05-06
#### Fixes
* Correct check hash [`#26638`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/46)

### v0.1.130 -  2022-04-05
#### Fixes
* Removed methods from readme [`#24721`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/45)

### v0.1.129 -  2022-03-28
#### Features
* New method to block addresses, and get blocked addresses list [`#24721`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/44)

### v0.1.128 -  2022-03-31
#### Fixes
* Remove memo log [`PR`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/43)

### v0.1.126 -  2022-03-18
#### Features
* Validation for memo field in transactions [`#24607`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/41)

### v0.1.125 -  2022-03-03
#### Features
* Error handlers for get and save wallets methods [`PR`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/40)

### v0.1.124 -  2022-02-28
#### Features
* Generated wallets `#21631`
  * Field for generated wallets [`PR`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/31)
  * Method for ID [`PR`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/33)
  * Method generateAndSwitchAccount [`PR`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/35)
  * JSON.stringify in method sendAndSaveGeneratedWallets [`PR`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/38)

#### Fixes
* Hotfix for swap redeem [`PR`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/39)

### v0.1.112 - 2022-02-16
#### Features
* Id for mint nft transaction [`#22100`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/37)
* Get nft by hash method to api [`#22259`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/36)
* The ability to set custom nonce for transactions [`#19886`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/28)

### v0.1.104 - 2021-11-30
#### Features
* Mechanism for processing a nft list request, taking into account private [`#17374`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/20)

### v0.1.101 - 2021-10-29
#### Features
* Mechanism for determining the address from the signature on the backend-console and returning full information on request if the address was ever owned by nft [`#12482`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/16)
* New transaction "nft reserve increase" [`#14007`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/13)
* Implement a check for current ownership of the nft address decimal [`#13072`](https://bitbucket.org/decimalteam/decimal-js-sdk/pull-requests/8)

src/tx.js
import DecimalNumber from 'decimal.js';
import TX_TYPE from './txTypes';
import TX_BROADCAST_MODES from './txBroadcastModes';
import ACCOUNT_INFO_MODES from './accountInfoModes';
import validateTxData from './validator';
import { formTx, prepareTx, postTx } from './txUtils';
import { getAmountFromUNI, getAmountToUNI } from './math';
import { redeemCheck } from './check';
import { getCommission } from './fees';
import { verifyAddress, verifyCheck } from './utils';

// constants
const MAX_MEMO_BYTES_LENGTH = 256;

DecimalNumber.set({ precision: 40 });
function sendCoinData(data, wallet) {
  return {
    sender: wallet.address,
    receiver: data.to,
    coin: {
      amount: getAmountToUNI(data.amount),
      denom: data.coin.toLowerCase(),
    },
  };
}

function burnCoinData(data, wallet) {
  return {
    sender: wallet.address,
    coin: {
      amount: getAmountToUNI(data.amount),
      denom: data.coin.toLowerCase(),
    },
  };
}

function buyCoinData(data, wallet) {
  const maxSpendLimit = data.maxSpendLimit ? getAmountToUNI(data.maxSpendLimit) : getAmountToUNI('100000000000');
  return {
    sender: wallet.address,
    coin_to_buy: {
      amount: getAmountToUNI(data.amount),
      denom: data.buyCoin.toLowerCase(),
    },
    max_coin_to_sell: {
      amount: maxSpendLimit,
      denom: data.spendCoin.toLowerCase(),
    },
  };
}

function sellCoinData(data, wallet) {
  const minBuyLimit = data.minBuyLimit ? getAmountToUNI(data.minBuyLimit) : '1';
  return {
    sender: wallet.address,
    coin_to_sell: {
      amount: getAmountToUNI(data.amount),
      denom: data.sellCoin.toLowerCase(),
    },
    min_coin_to_buy: {
      amount: minBuyLimit,
      denom: data.getCoin.toLowerCase(),
    },
  };
}

function sellAllCoinsData(data, wallet) {
  const minBuyLimit = data.minBuyLimit ? getAmountToUNI(data.minBuyLimit) : '1';
  return {
    sender: wallet.address,
    coin_to_sell: {
      amount: '0',
      denom: data.sellCoin.toLowerCase(),
    },
    min_coin_to_buy: {
      amount: minBuyLimit,
      denom: data.getCoin.toLowerCase(),
    },
  };
}

function delegate(data, wallet) {
  return {
    delegator_address: wallet.address,
    validator_address: data.address,
    coin: {
      denom: data.coin.toLowerCase(),
      amount: getAmountToUNI(data.stake),
    },
  };
}

function unbond(data, wallet) {
  return {
    delegator_address: wallet.address,
    validator_address: data.address,
    coin: {
      denom: data.coin.toLowerCase(),
      amount: getAmountToUNI(data.stake),
    },
  };
}

function declareCandidate(data, wallet) {
  return {
    commission: new DecimalNumber(data.commission).div(100).toFixed(18),
    validator_addr: wallet.validatorAddress,
    reward_addr: data.rewardAddress,
    pub_key: {
      type: 'tendermint/PubKeyEd25519',
      value: data.pubKey,
    },
    stake: {
      denom: data.coin.toLowerCase(),
      amount: getAmountToUNI(data.stake),
    },
    description: {
      moniker: data.description.moniker,
      identity: data.description.identity,
      website: data.description.website,
      security_contact: data.description.securityContact,
      details: data.description.details,
    },
  };
}

function editCandidate(data, wallet) {
  return {
    validator_address: wallet.validatorAddress,
    reward_address: data.rewardAddress,
    description: {
      moniker: data.description.moniker,
      identity: data.description.identity,
      website: data.description.website,
      security_contact: data.description.securityContact,
      details: data.description.details,
    },
  };
}

function disableEnableValidator(wallet) {
  return {
    validator_address: wallet.validatorAddress,
  };
}

function createCoin(data, wallet) {
  return {
    sender: wallet.address,
    title: data.title,
    symbol: data.ticker,
    constant_reserve_ratio: data.crr,
    initial_volume: getAmountToUNI(data.initSupply),
    initial_reserve: getAmountToUNI(data.reserve),
    limit_volume: getAmountToUNI(data.maxSupply),
    identity: data.identity,
  };
}

function updateCoin(data, wallet) {
  return {
    sender: wallet.address,
    symbol: data.ticker,
    identity: data.identity,
    limit_volume: getAmountToUNI(data.maxSupply),
  };
}

function multisigCreate(data, wallet) {
  return {
    sender: wallet.address,
    owners: data.owners,
    weights: data.weights,
    threshold: data.threshold,
  };
}

function multisigCreateTx(data, wallet) {
  return {
    sender: wallet.address,
    wallet: data.from,
    receiver: data.to,
    coins: [
      {
        denom: data.coin.toLowerCase(),
        amount: getAmountToUNI(data.amount),
      },
    ],
  };
}
function multisigSignTx(data, wallet) {
  return {
    sender: wallet.address,
    tx_id: data.txId,
  };
}

function multisend(data, wallet) {
  const obj = {
    sender: wallet.address,
    sends: [],
  };

  data.forEach((item) => {
    obj.sends.push({
      receiver: item.to,
      coin: {
        amount: getAmountToUNI(item.amount),
        denom: item.coin.toLowerCase(),
      },
    });
  });

  return obj;
}

function submitProposal(data, wallet) {
  if (!wallet.availableProposalSubmit) {
    throw new Error('This address is not available for this transaction');
  }

  return {
    content: data.content,
    proposer: wallet.address,
    voting_start_block: data.startBlock,
    voting_end_block: data.endBlock,
  };
}

function voteProposal(data, wallet) {
  return {
    proposal_id: data.id,
    voter: wallet.validatorAddress,
    option: data.decision,
  };
}

function swapInit(data, wallet) {
  return {
    from: wallet.address,
    recipient: data.recipient,
    amount: getAmountToUNI(data.amount),
    token_symbol: data.tokenSymbol,
    transaction_number: Date.now().toString(),
    from_chain: '1',
    dest_chain: data.destChain,
  };
}
function swapRedeem(data, wallet) {
  return {
    sender: wallet.address,
    from: data.from,
    recipient: data.recipient,
    amount: getAmountToUNI(data.amount),
    token_symbol: data.tokenSymbol,
    transaction_number: data.transactionNumber,
    from_chain: data.fromChain,
    dest_chain: '1',
    v: data.v,
    r: data.r.slice(2),
    s: data.s.slice(2),
  };
}

function nftMint(data, wallet) {
  return {
    denom: data.denom,
    id: data.id,
    sender: wallet.address,
    recipient: data.recipient ? data.recipient : wallet.address,
    quantity: data.quantity,
    reserve: getAmountToUNI(data.reserve),
    token_uri: data.token_uri,
    allow_mint: data.allow_mint,
  };
}
function nftBurn(data, wallet) {
  return {
    sender: wallet.address,
    denom: data.denom,
    id: data.id,
    sub_token_ids: data.sub_token_ids,
  };
}
function nftEditMetadata(data, wallet) {
  return {
    sender: wallet.address,
    denom: data.denom,
    id: data.id,
    token_uri: data.token_uri,
  };
}
function nftTransfer(data, wallet) {
  return {
    denom: data.denom,
    id: data.id,
    sender: wallet.address,
    recipient: data.recipient,
    sub_token_ids: data.sub_token_ids,
  };
}

function nftDelegate(data, wallet) {
  return {
    denom: data.denom,
    id: data.id,
    delegator_address: wallet.address,
    validator_address: data.validator_address,
    sub_token_ids: data.sub_token_ids,
  };
}
function nftUnbond(data, wallet) {
  return {
    denom: data.denom,
    id: data.id,
    delegator_address: wallet.address,
    validator_address: data.validator_address,
    sub_token_ids: data.sub_token_ids,
  };
}

function nftUpdateReserve(data, wallet) {
  return {
    sender: wallet.address,
    id: data.id,
    denom: data.denom,
    sub_token_ids: data.sub_token_ids,
    reserve: getAmountToUNI(data.reserve),
  };
}

function getValue(type, data, options, wallet) {
  validateTxData(data, type);

  if (options && options.feeCoin) {
    options.feeCoin = options.feeCoin.toLowerCase();
  }

  if (options && options.nonce) {
    if (typeof options.nonce !== 'string') {
      throw new Error('Custom nonce should be a string');
    }

    if (Number.isNaN(parseInt(options.nonce, 10))) {
      throw new Error('Custom nonce should be valid number string');
    }
  }

  if (options && options.txBroadcastMode) {
    if (!Object.values(TX_BROADCAST_MODES).includes(options.txBroadcastMode)) {
      throw new Error('Tx broadcast mode should be one of TX_BROADCAST_MODES');
    }
  }

  if (options && options.accountInfoMode) {
    if (!Object.values(ACCOUNT_INFO_MODES).includes(options.accountInfoMode)) {
      throw new Error('Account info mode should be one of ACCOUNT_INFO_MODES');
    }
  }

  if (options && options.setNonceAutomatically) {
    if (typeof options.setNonceAutomatically !== 'boolean') {
      throw new Error('Set nonce automatically should be a boolean');
    }
  }

  if (options && options.memo) {
    if (typeof options.memo !== 'string' || (new TextEncoder().encode(options.memo)).length > MAX_MEMO_BYTES_LENGTH) {
      throw new Error(`Memo should be a string with maximum ${MAX_MEMO_BYTES_LENGTH} bytes length`);
    }
  }

  let value = {};
  switch (type) {
    case TX_TYPE.COIN_SEND:
      value = sendCoinData(data, wallet);
      break;
    case TX_TYPE.COIN_BURN:
      value = burnCoinData(data, wallet);
      break;
    case TX_TYPE.COIN_BUY:
      value = buyCoinData(data, wallet);
      break;
    case TX_TYPE.COIN_SELL:
      value = sellCoinData(data, wallet);
      break;
    case TX_TYPE.COIN_SELL_ALL:
      value = sellAllCoinsData(data, wallet);
      break;
    case TX_TYPE.VALIDATOR_DELEGATE:
      value = delegate(data, wallet);
      break;
    case TX_TYPE.VALIDATOR_UNBOND:
      value = unbond(data, wallet);
      break;
    case TX_TYPE.VALIDATOR_CANDIDATE:
      value = declareCandidate(data, wallet);
      break;
    case TX_TYPE.VALIDATOR_CANDIDATE_EDIT:
      value = editCandidate(data, wallet);
      break;
    case TX_TYPE.VALIDATOR_SET_ONLINE:
    case TX_TYPE.VALIDATOR_SET_OFFLINE:
      value = disableEnableValidator(wallet);
      options = data;
      break;
    case TX_TYPE.COIN_CREATE:
      value = createCoin(data, wallet);
      break;
    case TX_TYPE.COIN_UPDATE:
      value = updateCoin(data, wallet);
      break;
    case TX_TYPE.COIN_REDEEM_CHECK:
      value = redeemCheck(data, wallet);
      break;
    case TX_TYPE.MULTISIG_CREATE_WALLET:
      value = multisigCreate(data, wallet);
      break;
    case TX_TYPE.MULTISIG_CREATE_TX:
      value = multisigCreateTx(data, wallet);
      break;
    case TX_TYPE.MULTISIG_SIGN_TX:
      value = multisigSignTx(data, wallet);
      break;
    case TX_TYPE.COIN_MULTISEND:
      value = multisend(data, wallet);
      break;
    case TX_TYPE.PROPOSAL_SUBMIT:
      value = submitProposal(data, wallet);
      break;
    case TX_TYPE.PROPOSAL_VOTE:
      value = voteProposal(data, wallet);
      break;
    case TX_TYPE.SWAP_INIT:
      value = swapInit(data, wallet);
      break;
    case TX_TYPE.SWAP_REDEEM:
      value = swapRedeem(data, wallet);
      break;
    case TX_TYPE.NFT_MINT:
      value = nftMint(data, wallet);
      break;
    case TX_TYPE.NFT_BURN:
      value = nftBurn(data, wallet);
      break;
    case TX_TYPE.NFT_EDIT_METADATA:
      value = nftEditMetadata(data, wallet);
      break;
    case TX_TYPE.NFT_TRANSFER:
      value = nftTransfer(data, wallet);
      break;
    case TX_TYPE.NFT_DELEGATE:
      value = nftDelegate(data, wallet);
      break;
    case TX_TYPE.NFT_UNBOND:
      value = nftUnbond(data, wallet);
      break;
    case TX_TYPE.NFT_UPDATE_RESERVE:
      value = nftUpdateReserve(data, wallet);
      break;
    default:
      throw new Error('Invalid type of transaction');
  }

  return { value, options };
}

export function getTransaction(api, wallet, decimal) {
  return async (type, data, options) => {
    const formatted = getValue(type, data, options, wallet);
    const broadcastTx = await formTx(api, wallet, decimal)(type, formatted.value, formatted.options);
    console.log('broadcast TX: ', broadcastTx);
    return broadcastTx;
  };
}

function validateData(data) {
  if (
    data.from === ''
    || data.sender === ''
    || data.to === ''
    || data.recipient === ''
    || data.validator_address === ''
    || data.delegator_address === ''
  ) {
    throw new Error('Address cannot be empty string');
  }

  if (data.check) {
    if (!verifyCheck(data.check)) {
      throw new Error('Incorrect check format');
    }
  }

  if (data.from) {
    if (!verifyAddress(data.from)) {
      throw new Error('Incorrect sender address format');
    }
  }

  if (data.sender) {
    if (!verifyAddress(data.sender)) {
      throw new Error('Incorrect sender address format');
    }
  }

  if (data.to) {
    if (!verifyAddress(data.to)) {
      throw new Error('Incorrect recipient address format');
    }
  }

  if (data.recipient) {
    if (!verifyAddress(data.recipient)) {
      throw new Error('Incorrect recipient address format');
    }
  }

  if (data.owners) {
    data.owners.forEach((address) => {
      if (!verifyAddress(address)) {
        throw new Error('Incorrect owner address format');
      }
    });
  }

  if (data.validator_address) {
    if (!verifyAddress(data.validator_address, 'dxvaloper')) {
      throw new Error('Incorrect validator address format');
    }
  }

  if (data.delegator_address) {
    if (!verifyAddress(data.delegator_address)) {
      throw new Error('Incorrect delegator address format');
    }
  }

  if (data.sends) {
    data.sends.forEach(({ receiver }) => {
      if (!verifyAddress(receiver)) {
        throw new Error('Incorrect owner address format');
      }
    });
  }
}

export function sendTransaction(type, api, wallet, decimal) {
  return async (data, options) => {
    /* FIXME: swap contains external blockchain address
     * format that needs to be verified differently.
     */
    if (type !== TX_TYPE.SWAP_INIT && type !== TX_TYPE.SWAP_REDEEM) {
      validateData(data);
    }

    const broadcastTx = await getTransaction(api, wallet, decimal)(type, data, options);
    console.log(broadcastTx)
    const txResult = await postTx(api, wallet)(broadcastTx);

    return txResult;
  };
}

export function estimateTxFee(api, wallet, decimal) {
  return async (type, data, options) => {
    try {
      const { feeCoin } = options;

      if (feeCoin) {
        const broadcastTx = await getTransaction(api, wallet, decimal)(type, data, options);
        const feeAmounts = broadcastTx.tx.fee.amount;
        const fee = feeAmounts.length ? feeAmounts[0].amount : '0';

        return getAmountFromUNI(fee);
      }

      const formatted = getValue(type, data, options, wallet);
      const tx = await prepareTx(api)(type, formatted.value, formatted.options);
      const fee = await getCommission(api)(tx, 'del');

      return fee.value.times(0.001).toFixed();
    } catch (e) {
      console.log(e);
      return new Error('error');
    }
  };
}
